controllers/operatorpolicy_controller.go

```diff
@@ -1198,22 +1216,96 @@ func (r *OperatorPolicyReconciler) considerResolutionFailed(
 	}
 
 	if !includesSubscription {
+		if policy.Status.SubscriptionInterventionExpired() {
+			policy.Status.SubscriptionInterventionTime = nil
+		}
+
 		return mergedSub, nil, false, nil
 	}
 
 	// Handle non-ConstraintsNotSatisfiable reasons separately
 	if !strings.EqualFold(subResFailed.Reason, "ConstraintsNotSatisfiable") {
-		changed := updateStatus(policy, subResFailedCond(subResFailed), nonCompObj(foundSub, subResFailed.Reason))
+		changed := updateStatus(policy, subResFailedCond(subResFailed), nonCompObj(mergedSub, subResFailed.Reason))
+
+		if policy.Status.SubscriptionInterventionExpired() {
+			policy.Status.SubscriptionInterventionTime = nil
+		}
 
 		return mergedSub, nil, changed, nil
 	}
 
 	// A "constraints not satisfiable" message has nondeterministic clauses, and can be noisy with a list of versions.
 	// Just report a generic condition, which will prevent the OperatorPolicy status from constantly updating
 	// when the details in the Subscription status change.
-	changed := updateStatus(policy, subConstraintsNotSatisfiableCond, nonCompObj(foundSub, "ConstraintsNotSatisfiable"))
+	changed := updateStatus(policy, subConstraintsNotSatisfiableCond,
+		nonCompObj(mergedSub, "ConstraintsNotSatisfiable"))
+
+	if policy.Status.SubscriptionInterventionExpired() || policy.Status.SubscriptionInterventionTime == nil {
+		interventionTime := metav1.Time{Time: time.Now().Add(olmGracePeriod)}
+		policy.Status.SubscriptionInterventionTime = &interventionTime
+
+		opLog.V(1).Info("Detected ConstraintsNotSatisfiable, setting an intervention time",
+			"interventionTime", interventionTime, "subscription", mergedSub)
+
+		return mergedSub, nil, changed, nil
+	}
+
+	if policy.Status.SubscriptionInterventionWaiting() {
+		opLog.V(1).Info("Detected ConstraintsNotSatisfiable, giving OLM more time before possibly intervening",
+			"interventionTime", policy.Status.SubscriptionInterventionTime)
+
+		return mergedSub, nil, changed, nil
+	}
+
+	// Do the "intervention"
 
-	return mergedSub, nil, changed, nil
+	relatedCSVs := policy.Status.RelatedObjsOfKind("ClusterServiceVersion")
+	if len(relatedCSVs) != 1 {
+		opLog.V(1).Info("No related CSVs to use for an intervention")
+
+		return mergedSub, nil, changed, nil
+	}
+
+	csvIdx := 0
```

Comment:
Wouldn't this always be 0 because of `if len(relatedCSVs) != 1 {`?

Reply:
The `RelatedObjsOfKind` gives back a map :sweat_smile: I thought it would be a good idea because then it's easy to update the condition in-place... I mostly stand by that, but it does make this usage look odd.

Usually index 0 is the CatalogSource.